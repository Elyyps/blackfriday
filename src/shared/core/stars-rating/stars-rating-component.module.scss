.stars-rating {
  display: flex;
  &__star {
    margin-right: 4px;
  }
}
.full {
  path {
    fill: #ffbd03;
  }
}

.empty {
  path {
    fill: #fff;
  }
}
