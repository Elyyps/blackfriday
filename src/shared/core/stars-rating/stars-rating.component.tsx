import * as React from "react";

import styles from "./stars-rating-component.module.scss";
import Star from "@assets/icons/star-filled.svg";
import HalfStar from "@assets/icons/half-star-color.svg";
import { IconComponent } from "@app/core/";

export interface IStarsRatingComponentProps {
  numberOfStars?: number;
  rating: number;
  showRating?: boolean;
  voteTrigger?: () => void;
}

enum StarsType {
  empty = "EMPTY",
  full = "FULL",
  half = "HALF"
}
const defaultStars = 5;
const StarsRatingComponent = (props: IStarsRatingComponentProps) => {
  const { showRating, voteTrigger } = props;
  const [stars, setStars] = React.useState<StarsType[]>([]);
  const [currentHoverStar, setCurrentHoverStar] = React.useState<number>(-1);
  const totalStars = props.numberOfStars || defaultStars;
  const rating = props.rating > totalStars ? 0 : props.rating;
  const defineStars = () => {
    const finalStarsArray: StarsType[] = [];
    const integerPart = Math.trunc(rating);
    const hasHalfStar = rating - integerPart > 0;
    const emptyStars = totalStars - ((hasHalfStar ? 1 : 0) + integerPart);
    [...Array(Math.floor(integerPart))].map(e => finalStarsArray.push(StarsType.full));
    if (hasHalfStar) finalStarsArray.push(StarsType.half);
    [...Array(Math.floor(emptyStars))].map(e => finalStarsArray.push(StarsType.empty));
    setStars(finalStarsArray);
  };
  React.useEffect(() => {
    defineStars();
  }, []);

  return (
    <div
      style={{ cursor: voteTrigger ? "pointer" : "default" }}
      onClick={() => {
        if (voteTrigger) voteTrigger();
      }}
      role="button"
      onMouseLeave={() => setCurrentHoverStar(-1)}
      className={styles["stars-rating"]}
    >
      {showRating && <span className={styles["stars-rating__label"]}>{rating}</span>}
      {stars.map((star, i) => (
        <span
          onMouseOver={() => setCurrentHoverStar(i)}
          key={i}
          role="button"
          className={`${voteTrigger &&
            (currentHoverStar >= 0
              ? currentHoverStar >= i
                ? styles["stars-rating__hovered"]
                : styles["stars-rating__hovered-empty"]
              : "")} ${styles["stars-rating__star"]} ${
            styles[star === StarsType.full ? "full" : star === StarsType.half ? "half" : "empty"]
          }`}
        >
          <IconComponent
            icon={
              voteTrigger && currentHoverStar >= 0
                ? Star
                : star === StarsType.full
                ? Star
                : star === StarsType.half
                ? HalfStar
                : Star
            }
            size="14px"
          />
        </span>
      ))}
    </div>
  );
};

export { StarsRatingComponent };
