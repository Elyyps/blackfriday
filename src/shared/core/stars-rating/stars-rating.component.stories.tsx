import * as React from "react";
import { storiesOf } from "@storybook/react";
import { withA11y } from "@storybook/addon-a11y";
import { StarsRatingComponent } from "./stars-rating.component";
const numberStar = 5;
const rating = 1;

const notes = require("./stars-rating.md");

storiesOf("Stars-rating", module)
  .addDecorator(withA11y)
  .add("Basic implementation", () => <StarsRatingComponent numberStars={numberStar} rating={rating} />, {
    notes
  });
