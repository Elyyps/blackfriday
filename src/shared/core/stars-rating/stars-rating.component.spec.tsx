import React from "react";
import { shallow } from "enzyme";
import { StarsRatingComponent } from "./stars-rating.component";
const numberStar = 5;
const rating = 1;
describe("[StarsRating]", () => {
  it("should render component without crashing", () => {
    const renderedComponent = shallow(
      <StarsRatingComponent numberStars={numberStar} rating={rating} />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
