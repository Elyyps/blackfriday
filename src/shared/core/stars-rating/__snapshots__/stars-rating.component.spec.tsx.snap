// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[StarsRating] should render component without crashing 1`] = `
<div
  className="stars-rating"
/>
`;
